src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,509 @@
+	{
+		// track how many samples we've updated without a clock; if it's been more than 1/32 of a second, output silence
+		m_samples_generated += buffer.samples();
+		if (m_samples_generated > SAMPLE_RATE / 32)
+		{
+			m_samples_generated = SAMPLE_RATE;
+			m_next_sample = 0;
+		}
+	}
+*/
 
-	/* speech clock changing? */
-	if (is_active_clock_transition(clock_state))
+	if (is_external_oscillator())
 	{
-		/* update the output buffer before changing the registers */
-		m_channel->update();
+		/* external oscillator */
+		for (int i = 0; i < buffer.samples(); i++)
+		{
+			buffer.put(i, stream_buffer::sample_t(m_next_sample/32768.0));
 
-		/* clear the update count */
-		m_update_count = 0;
+			m_samples_generated++;
+
+			uint8_t clock_state = current_clock_state();
 
-		process_digit();
+			/* pull in next digit on the appropriate edge of the clock */
+			if (is_clock_changed(clock_state))
+			{
+				process_bit(m_buffered_bit, clock_state);
+			}
+
+			m_last_clock_state = clock_state;
+		}
 	}
 
-	/* update the clock */
-	m_last_clock_state = clock_state;
+	/* software driven clock */
+	else
+		for (int i = 0; i < buffer.samples(); i++)
+			buffer.put(i, stream_buffer::sample_t(m_next_sample/32768.0));
 }
 
 
-void hc55516_device::digit_w(int digit)
+
+//#########################################
+//                 HC55532
+//#########################################
+DEFINE_DEVICE_TYPE(HC55532, hc55532_device, "hc55532", "HC-55532")
+
+hc55532_device::hc55532_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: hc55516_device(mconfig, HC55532, tag, owner, clock, 0xf80, 7, 0xfe1, 5)
 {
-	if (is_external_oscillator())
-	{
-		m_channel->update();
-		m_new_digit = digit & 1;
-	}
-	else
-		m_digit = digit & 1;
 }
 
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
 
-int hc55516_device::clock_state_r()
+void hc55532_device::device_reset()
 {
-	/* only makes sense for setups with an external oscillator */
-	assert(is_external_oscillator());
+	// simulate /FZ having been held for a while
+	m_sylfilter = 0x7f;
+	m_intfilter = 0;
+	m_agc = true;
+	m_buffered_fzq = true; // assuming /FZ was just released and is now high/inactive
+}
 
-	m_channel->update();
 
-	return current_clock_state();
+
+//##########################################
+//                 MC3417
+//##########################################
+DEFINE_DEVICE_TYPE(MC3417, mc3417_device, "mc3417", "MC3417")
+
+mc3417_device::mc3417_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: cvsd_device_base(mconfig, MC3417, tag, owner, clock, FALLING, 0x7)
+	, m_charge(pow(exp(-1.0), 1.0 / (FILTER_CHARGE_TC * 16000.0)))
+	, m_decay(pow(exp(-1.0), 1.0 / (FILTER_DECAY_TC * 16000.0)))
+	, m_leak(pow(exp(-1.0), 1.0 / (INTEGRATOR_LEAK_TC * 16000.0)))
+	, m_sylfilter_d(0.0)
+	, m_intfilter_d(0.0)
+{
```

Comment:
Please use constructor delegation to avoid repetition here.

Reply:
done.